mutable_buffer/src/persistence_windows.rs

```diff
@@ -0,0 +1,602 @@
+
+const CLOSED_WINDOW_SECONDS: u8 = 30;
+const SECONDS_IN_MINUTE: u8 = 60;
+
+/// PersistenceWindows keep track of ingested data within a partition to determine when it
+/// can be persisted. This allows IOx to receive out of order writes (in their timestamps) while
+/// persisting mostly in non-time overlapping Parquet files.
+#[derive(Debug, Clone)]
+pub struct PersistenceWindows {
+    persistable: Option<Window>,
+    closed: Vec<Window>,
+    open: Option<Window>,
+    late_arrival_period: Duration,
+}
+
+impl PersistenceWindows {
+    pub fn new(late_arrival_minutes: u8) -> Self {
+        let late_arrival_seconds = late_arrival_minutes * SECONDS_IN_MINUTE;
+        let closed_window_count = late_arrival_seconds / CLOSED_WINDOW_SECONDS;
+
+        Self {
+            persistable: None,
+            closed: Vec::with_capacity(closed_window_count as usize),
+            open: None,
+            late_arrival_period: Duration::new(late_arrival_seconds as u64, 0),
+        }
+    }
+
+    /// Updates the windows with the information from a batch of rows to the same partition.
+    pub fn add_range(
+        &mut self,
+        sequencer_id: u32,
+        sequence_number: u64,
+        row_count: usize,
+        min_time: DateTime<Utc>,
+        max_time: DateTime<Utc>,
+        received_at: Instant,
+    ) {
+        self.rotate(received_at);
+
+        match self.open.as_mut() {
+            Some(w) => w.add_range(sequencer_id, sequence_number, row_count, min_time, max_time),
+            None => {
+                self.open = Some(Window::new(
+                    received_at,
+                    sequencer_id,
+                    sequence_number,
+                    row_count,
+                    min_time,
+                    max_time,
+                ))
+            }
+        };
+    }
+
+    /// Returns the number of rows that are persistable. These rows could be duplicates and there
+    /// are other rows that may fall in closed and open that would be pulled into a persistence
+    /// operation. This number is used to determine if persistence should be triggered, not as
+    /// an exact number.
+    pub fn persistable_row_count(&self) -> usize {
+        self.persistable.as_ref().map(|w| w.row_count).unwrap_or(0)
+    }
+
+    /// Returns the instant of the oldest persistable data. This is used by the lifecycle manager
+    /// to determine if persistence should be triggered because data has been sitting in memory
+    /// too long. This limit should only be hit if the throughput in the partition hasn't crossed
+    /// the row count threshold or if the partition has gone cold for writes.
+    pub fn persistable_age(&self) -> Option<Instant> {
+        self.persistable.as_ref().map(|w| w.created_at)
+    }
+
+    /// Returns the max timestamp of data in the persistable window. Any unpersisted data with a
+    /// timestamp <= than this value can be persisted.
+    pub fn t_flush(&self) -> Option<DateTime<Utc>> {
```

Comment:
```suggestion
    pub fn max_persistable_timestamp(&self) -> Option<DateTime<Utc>> {
```

Maybe for explicitness?